File: common/src/main/java/io/github/notenoughupdates/moulconfig/common/DynamicTextureReference.java
package io.github.notenoughupdates.moulconfig.common;

import org.jetbrains.annotations.ApiStatus;

import java.awt.image.BufferedImage;
import java.io.Closeable;

/**
 * A dynamically loaded texture. Must be destroyed manually.
 */
@ApiStatus.NonExtendable
public abstract class DynamicTextureReference implements Closeable {
    private boolean wasDestroyed;

    /**
     * An opaque reference to this dynamic texture. Can be used with {@link RenderContext#drawTexturedRect}.
     */
    public abstract MyResourceLocation getIdentifier();

    /**
     * Destroy this texture. Using {@link #getIdentifier()} after calling this will cause issues.
     */
    public final void destroy() {
        if (!wasDestroyed) {
            doDestroy();
        }
        wasDestroyed = true;
    }

    public abstract void update(BufferedImage bufferedImage);

    protected abstract void doDestroy();

    @Override
    public final void close() {
        destroy();
    }

    @Override
    protected void finalize() throws Throwable {
        try {
            if (!wasDestroyed) {
                IMinecraft.INSTANCE.getLogger("DynamicTextureReference").warn("Dangling DynamicTextureReference");
            }
        } finally {
            super.finalize();
        }
    }
}

File: common/src/main/java/io/github/notenoughupdates/moulconfig/common/DynamicTextureReference.kt
package io.github.notenoughupdates.moulconfig.common

import org.jetbrains.annotations.ApiStatus
import java.awt.image.BufferedImage
import java.io.Closeable

/**
 * A dynamically loaded texture. Must be destroyed manually.
 */
@ApiStatus.NonExtendable
abstract class DynamicTextureReference : Closeable {
    /**
     * An opaque reference to this dynamic texture. Can be used with [RenderContext.drawTexturedRect].
     */
    abstract val identifier: MyResourceLocation

    /**
     * Destroy this texture. Using [identifier] after calling this will cause issues.
     */
    fun destroy() {
        if (!wasDestroyed)
            doDestroy()
        wasDestroyed = true
    }

    abstract fun update(bufferedImage: BufferedImage) // TODO: check why this is broken on 1.21

    protected abstract fun doDestroy()

    private var wasDestroyed = false

    override fun close() {
        destroy()
    }

    protected fun finalize() { // TODO: replace with a reference queue and Warnings.warn
        if (!wasDestroyed)
            IMinecraft.instance.getLogger("DynamicTextureReference").warn("Dangling DynamicTextureReference")
    }
}
